### PR TITLE
 feat: add support for jsonldContext member in subscriptions

### PR DESCRIPTION
**Pull request by ranim-n**

### CODE REVIEW

**Comment by bobeal** [comment]
some points are missing in the implementation.

for instance, in 5.8.1.4, it is said: `The @context to be used for sending Notifications related to this Subscription shall be the one specified in the jsonldContext field. If not present, the jsonldContext field shall be initialized with the @context applicable for the Subscription (if @context is also not present in the Subscription, see clause 5.5.5). When the remote JSON-LD @context referenced by the jsonldContext field is not available implementations shall raise an error of type LdContextNotAvailable. If the remote JSON-LD @context referenced by the jsonldContext field is invalid, implementations shall raise an error of type BadRequestData.`

(search for `jsonldContext` in the specification)

**Comment by ranim-n** [comment]
> some points are missing in the implementation.
> 
> for instance, in 5.8.1.4, it is said: `The @context to be used for sending Notifications related to this Subscription shall be the one specified in the jsonldContext field. If not present, the jsonldContext field shall be initialized with the @context applicable for the Subscription (if @context is also not present in the Subscription, see clause 5.5.5). When the remote JSON-LD @context referenced by the jsonldContext field is not available implementations shall raise an error of type LdContextNotAvailable. If the remote JSON-LD @context referenced by the jsonldContext field is invalid, implementations shall raise an error of type BadRequestData.`
> 
> (search for `jsonldContext` in the specification)

to check if a jsonld context is available or valid do I make an http request to it to make sure it is reachable or do I use it in compaction/expansion to test it... ?

**Comment by bobeal** [comment]
> to check if a jsonld context is available or valid do I make an http request to it to make sure it is reachable or do I use it in compaction/expansion to test it... ?

Doing a compaction or expansion is too expensive. So option 1.

The JSON-LD library should already have some functions to do it (this is something it is doing every time we expand / compact). Maybe in https://github.com/filip26/titanium-json-ld/blob/main/src/main/java/com/apicatalog/jsonld/loader/DefaultHttpLoader.java?

**Comment by ranim-n** [comment]
> > to check if a jsonld context is available or valid do I make an http request to it to make sure it is reachable or do I use it in compaction/expansion to test it... ?
> 
> Doing a compaction or expansion is too expensive. So option 1.
> 
> The JSON-LD library should already have some functions to do it (this is something it is doing every time we expand / compact). Maybe in https://github.com/filip26/titanium-json-ld/blob/main/src/main/java/com/apicatalog/jsonld/loader/DefaultHttpLoader.java?

when updating a subscription there are no checks (see https://github.com/stellio-hub/stellio-context-broker/issues/1090 ) should I do a check for the` jsonldContext`  as stated in 5.8.2.4 anyway ?

**Comment by bobeal** [comment]
> > > to check if a jsonld context is available or valid do I make an http request to it to make sure it is reachable or do I use it in compaction/expansion to test it... ?
> > 
> > 
> > Doing a compaction or expansion is too expensive. So option 1.
> > The JSON-LD library should already have some functions to do it (this is something it is doing every time we expand / compact). Maybe in https://github.com/filip26/titanium-json-ld/blob/main/src/main/java/com/apicatalog/jsonld/loader/DefaultHttpLoader.java?
> 
> when updating a subscription there are no checks (see #1090 ) should I do a check for the` jsonldContext` as stated in 5.8.2.4 anyway ?

Start with the creation only. Will see later.

File: subscription-service/config/detekt/baseline.xml [hunk 1]
```diff
@@ -2,10 +2,17 @@
...
     <ID>LongParameterList:FixtureUtils.kt$( withQueryAndGeoQuery: Pair&lt;Boolean, Boolean&gt; = Pair(true, true), withEndpointReceiverInfo: Boolean = true, withNotifParams: Pair&lt;FormatType, List&lt;String&gt;&gt; = Pair(FormatType.NORMALIZED, emptyList()), withModifiedAt: Boolean = false, georel: String = "within", geometry: String = "Polygon", coordinates: String = "[[[100.0, 0.0], [101.0, 0.0], [101.0, 1.0], [100.0, 1.0], [100.0, 0.0]]]", timeInterval: Int? = null, contexts: List&lt;String&gt; = listOf(NGSILD_TEST_CORE_CONTEXT) )</ID>
+    <ID>MaxLineLength:NotificationServiceTests.kt$NotificationServiceTests$fun</ID>
+    <ID>MaxLineLength:SubscriptionServiceTests.kt$SubscriptionServiceTests$val subscription = ParsingUtils.parseSubscription(rawSubscription.deserializeAsMap(), emptyList()).shouldSucceedAndResult()</ID>
+    <ID>MaximumLineLength:NotificationServiceTests.kt$NotificationServiceTests$ </ID>
+    <ID>MaximumLineLength:SubscriptionServiceTests.kt$SubscriptionServiceTests$ </ID>
+    <ID>NoUnusedImports:SubscriptionService.kt$com.egm.stellio.subscription.service.SubscriptionService.kt</ID>
```
Comment (thomasBousselin) [inline subscription-service/config/detekt/baseline.xml:15]: You should avoid adding to the baseline if it's not necessary.

File: shared/config/detekt/baseline.xml [hunk 2]
```diff
@@ -7,8 +7,12 @@
...
     <ID>LongParameterList:ApiResponses.kt$( entities: Any, count: Int, resourceUrl: String, paginationQuery: PaginationQuery, requestParams: MultiValueMap&lt;String, String&gt;, mediaType: MediaType, contexts: List&lt;String&gt; )</ID>
+    <ID>MaxLineLength:ApiExceptions.kt$if</ID>
+    <ID>MaximumLineLength:ApiExceptions.kt$ </ID>
     <ID>SpreadOperator:EntityEvent.kt$EntityEvent$( *[ JsonSubTypes.Type(value = EntityCreateEvent::class), JsonSubTypes.Type(value = EntityReplaceEvent::class), JsonSubTypes.Type(value = EntityDeleteEvent::class), JsonSubTypes.Type(value = AttributeAppendEvent::class), JsonSubTypes.Type(value = AttributeReplaceEvent::class), JsonSubTypes.Type(value = AttributeUpdateEvent::class), JsonSubTypes.Type(value = AttributeDeleteEvent::class), JsonSubTypes.Type(value = AttributeDeleteAllInstancesEvent::class) ] )</ID>
+    <ID>SwallowedException:JsonLdUtils.kt$JsonLdUtils$e: Exception</ID>
     <ID>SwallowedException:JsonLdUtils.kt$JsonLdUtils$e: JsonLdError</ID>
+    <ID>TooGenericExceptionCaught:JsonLdUtils.kt$JsonLdUtils$e: Exception</ID>
```
Comment (thomasBousselin) [inline shared/config/detekt/baseline.xml:15]: You should avoid adding to the baseline if it's not necessary.

File: subscription-service/src/test/resources/ngsild/subscription_full.json [hunk 3]
```diff
@@ -38,5 +38,8 @@
   },
   "expiresAt": "2100-01-01T00:00:00Z",
   "throttling": 60,
-  "lang": "fr,en"
+  "lang": "fr,en",
+  "jsonldContext": [
+    "http://localhost:8093/jsonld-contexts/apic-compound.jsonld"
+  ]
```
Comment (bobeal) [inline subscription-service/src/test/resources/ngsild/subscription_full.json:44]: see 5.2.12: `jsonLdContext` is a string, not an array of strings

File: shared/src/main/kotlin/com/egm/stellio/shared/util/JsonLdUtils.kt [hunk 4]
```diff
@@ -244,6 +248,15 @@ object JsonLdUtils {
         }
     }
 
+    fun checkJsonldContext(context: URI) {
```
[the remark this answers is not in the document]
Comment (ranim-n) [reply shared/src/main/kotlin/com/egm/stellio/shared/util/JsonLdUtils.kt:251]: can you explain more ? How will the check go if no exception is thrown ?

[the remark this answers is not in the document]
Comment (bobeal) [reply shared/src/main/kotlin/com/egm/stellio/shared/util/JsonLdUtils.kt:251]: Either

File: subscription-service/src/test/kotlin/com/egm/stellio/subscription/service/NotificationServiceTests.kt [hunk 5]
```diff
@@ -215,6 +215,47 @@ class NotificationServiceTests {
...
+            ATTRIBUTE_UPDATED
+        ).shouldSucceedWith { notificationResults ->
+            val notificationResult = notificationResults[0]
+            assertEquals(subscription.id, notificationResult.first.id)
+            assertEquals(subscription.id, notificationResult.second.subscriptionId)
+            assertEquals(1, notificationResult.second.data.size)
+            assertTrue(notificationResult.second.data[0].containsKey(NGSILD_NAME_TERM))
+            assertTrue(notificationResult.second.data[0].containsKey(MANAGED_BY_COMPACT_RELATIONSHIP))
+            assertEquals(APIC_COMPOUND_CONTEXT, notificationResult.second.data[0][JsonLdUtils.JSONLD_CONTEXT])
+            assertTrue(notificationResult.third)
```
[the remark this answers is not in the document]
Comment (ranim-n) [reply subscription-service/src/test/kotlin/com/egm/stellio/subscription/service/NotificationServiceTests.kt:255]: The keys are not the same. In the previous entity they are  `NGSILD_NAME_PROPERTY` and `MANAGED_BY_RELATIONSHIP)` . In this entity they are `NGSILD_NAME_TERM` and `MANAGED_BY_COMPACT_RELATIONSHIP`

[the remark this answers is not in the document]
Comment (bobeal) [reply subscription-service/src/test/kotlin/com/egm/stellio/subscription/service/NotificationServiceTests.kt:255]: indeed, my bad

**Review by thomasBousselin — APPROVED** [review]

**Comment by bobeal** [comment]
@ranim-n don't forget the last comments in this PR

**Review by bobeal — APPROVED** [review]

File: subscription-service/src/main/kotlin/com/egm/stellio/subscription/service/SubscriptionService.kt [hunk 6]
```diff
@@ -269,25 +281,30 @@ class SubscriptionService(
...
-        } else
-            buildContextLinkHeader(subscription.contexts[0])
+    fun getContextsLink(subscription: Subscription): String {
+        val contextLink = when {
+            subscription.contexts.size > 1 && subscription.jsonldContext == null -> {
+                val linkToRetrieveContexts = subscriptionProperties.stellioUrl +
+                    "/ngsi-ld/v1/subscriptions/${subscription.id}/context"
+                linkToRetrieveContexts
+            }
+            subscription.jsonldContext != null -> subscription.jsonldContext.toString()
```
Comment (bobeal) [inline subscription-service/src/main/kotlin/com/egm/stellio/subscription/service/SubscriptionService.kt:303]: If you do this check first, you will not need the second condition in the previous case (better readability)

**Comment by bobeal** [comment]
LGTM

**Comment by ranim-n** [comment]
@thomasBousselin

**Review by thomasBousselin — APPROVED** [review]

**Pull request merged**